sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py

```diff
@@ -0,0 +1,629 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+import time
+import datetime
+import logging
+import functools
+import uuid
+from typing import Any, List, TYPE_CHECKING, Optional, Union
+import six
+
+from uamqp import ReceiveClient, Source, types
+from uamqp.constants import SenderSettleMode
+
+from ._base_handler import BaseHandler
+from ._common.utils import create_authentication
+from ._common.message import PeekMessage, ReceivedMessage
+from ._common.constants import (
+    REQUEST_RESPONSE_GET_SESSION_STATE_OPERATION,
+    REQUEST_RESPONSE_SET_SESSION_STATE_OPERATION,
+    REQUEST_RESPONSE_RENEW_SESSION_LOCK_OPERATION,
+    REQUEST_RESPONSE_RECEIVE_BY_SEQUENCE_NUMBER,
+    REQUEST_RESPONSE_UPDATE_DISPOSTION_OPERATION,
+    REQUEST_RESPONSE_RENEWLOCK_OPERATION,
+    REQUEST_RESPONSE_PEEK_OPERATION,
+    ReceiveSettleMode,
+    NEXT_AVAILABLE,
+    SESSION_LOCKED_UNTIL,
+    DATETIMEOFFSET_EPOCH,
+    SESSION_FILTER,
+    REQUEST_RESPONSE_RENEWLOCK_OPERATION
+)
+from .exceptions import (
+    _ServiceBusErrorPolicy,
+    SessionLockExpired
+)
+from ._common import mgmt_handlers
+
+if TYPE_CHECKING:
+    from azure.core.credentials import TokenCredential
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ServiceBusSession(object):
+    """
+    The ServiceBusSession is used for manage session states and lock renewal.
+
+    **Please use the instance variable `session` on the ServiceBusReceiver to get the corresponding ServiceBusSession
+    object linked with the receiver instead of instantiating a ServiceBusSession object directly.**
+
+    .. admonition:: Example:
+
+        .. literalinclude:: ../samples/sync_samples/sample_code_servicebus.py
+            :start-after: [START get_session_sync]
+            :end-before: [END get_session_sync]
+            :language: python
+            :dedent: 4
+            :caption: Get session from a receiver
+    """
+    def __init__(self, session_id, receiver, encoding="UTF-8"):
+        self._session_id = session_id
+        self._receiver = receiver
+        self._encoding = encoding
+        self._session_start = None
+        self._locked_until = None
+        self.auto_renew_error = None
+
+    def _can_run(self):
```

Comment:
as with _is_live, if _can_run doesn't actually answer a boolean question(can run?) we should change the naming to mirror the actual semantics.  I use "ensure" as a broad gimmic for situations like this, but that's my own silly pattern.  being explicit and naming closer to the functionality may be even nicer since this function seems quite targeted, "_validate_not_expired"?  (Unless you think this functionality will grow and become generic in the future, but we can always change it then if it does)

Reply:
https://github.com/Azure/azure-sdk-for-python/issues/10602